crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -279,66 +295,343 @@ where
+            &msg.message.packet.port_id_on_a,
+            &msg.message.packet.chan_id_on_a,
+            msg.message.packet.seq_on_a,
+        );
+        if !keys_changed.contains(&ack_key) {
+            // Otherwise ignore this event
+            return Ok(());
+        }
+        let acknowledgement = match self.ctx.read_bytes_post(&ack_key)? {
+            Some(value) => AcknowledgementCommitment::from(value),
+            None => return Ok(()),
+        };
+        // Get the pre-image of the acknowledgement commitment
+        let Some(acknowledgement) = ibc_acks.get(&acknowledgement.into_vec())
+        else {
+            return Ok(());
+        };
+        // If the transfer was a failure, then enable funds to
+        // be withdrawn from the IBC internal address
+        if acknowledgement.is_successful() {
+            // Mirror how the IBC token is derived in
+            // gen_ibc_shielded_transfer in the non-refund case
+            let ibc_denom = self.query_ibc_denom(
+                packet_data.token.denom.to_string(),
+                Some(&Address::Internal(InternalAddress::Ibc)),
+            )?;
+            let token = namada_ibc::received_ibc_token(
+                ibc_denom,
+                &msg.message.packet.port_id_on_a,
+                &msg.message.packet.chan_id_on_a,
+                &msg.message.packet.port_id_on_b,
+                &msg.message.packet.chan_id_on_b,
+            )
+            .into_storage_result()
+            .map_err(Error::NativeVpError)?;
+            let delta = ValueSum::from_pair(
+                token.clone(),
+                Amount::from_uint(Uint(*packet_data.token.amount), 0).unwrap(),
+            );
+            // Enable funds to be taken from the IBC internal
+            // address and be deposited elsewhere
+            // Required for the IBC internal Address to release
+            // funds
+            let pre_entry =
+                acc.pre.entry(addr_taddr(IBC)).or_insert(ValueSum::zero());
+            *pre_entry =
+                checked!(pre_entry + &delta).map_err(native_vp::Error::new)?;
+        }
+        Ok(())
+    }
+
+    // Apply relevant IBC packets to the changed balances structure
+    fn apply_ibc_packet(
+        &self,
+        mut acc: ChangedBalances,
+        ibc_msg: &IbcMessage,
+        keys_changed: &BTreeSet<Key>,
+        ibc_acks: &BTreeMap<Vec<u8>, AcknowledgementStatus>,
+    ) -> Result<ChangedBalances> {
+        match ibc_msg {
+            // This event is emitted on the sender
+            IbcMessage::Transfer(msg) => {
+                // Get the packet commitment from post-storage that corresponds
+                // to this event
+                let receiver = msg.message.packet_data.receiver.to_string();
+                let addr = TAddrData::Ibc(receiver.clone());
+                acc.decoder.insert(ibc_taddr(receiver), addr);
+                self.apply_send_packet(&mut acc, msg, keys_changed)?;
```

Comment:
We could generalize `apply_send_packet` for `Transfer` and `NftTransfer`

Reply:
Agreed. I was just a bit reluctant to do this now because the NFT E2E tests are not a part of the tests crate in the Namada repo and require some additional setup which I've not done yet. Do you think you could get this done more quickly/efficiently than me and open up a separate PR to generalize `Transfer` to `NftTransfer`? If not, then I can start working on this...

Reply:
I hope this could be helpful: https://github.com/anoma/namada/pull/3422

Reply:
Brilliant, thanks! Did you manage to run the NFT E2E tests on this? Just want to clarify https://github.com/anoma/namada/pull/3409#discussion_r1644380375 , then I'll merge this in.

Reply:
To be honest, I didn't try the NFT E2E test. I will test it after testing fungible token transfers in this PR.
Also, I'll release the updated Hermes for this PR to test in CI.

Reply:
Ah, I see thanks. I've partially applied #3422 where it pertains to determining a token `Address` from a packet `denom`. However I just wanted to clarify whether or not logic to receive NFT messages is missing before applying the rest of the PR. See https://github.com/anoma/namada/pull/3422/files#r1645836855 . Thanks.